sql-odbc/src/odfesqlodbc/odbcapiw.c

```diff
@@ -1005,3 +1006,63 @@ RETCODE SQL_API SQLSetConnectOptionW(HDBC ConnectionHandle, SQLUSMALLINT Option,
     LEAVE_CONN_CS(conn);
     return ret;
 }
+
+RETCODE SQL_API SQLErrorW(SQLHENV EnvironmentHandle, SQLHDBC ConnectionHandle,
+                          SQLHSTMT StatementHandle, SQLWCHAR *Sqlstate,
+                          SQLINTEGER *NativeError, SQLWCHAR *MessageText,
+                          SQLSMALLINT BufferLength, SQLSMALLINT *TextLength) {
+    RETCODE ret;
+    SQLSMALLINT buflen;
+    SQLSMALLINT tlen = 0;
+    SQLSMALLINT RecNumber = 1;
+    char qstr_ansi[8], *mtxt = NULL;
+
+    MYLOG(ES_TRACE, "entering\n");
+    buflen = 0;
+    if (MessageText && BufferLength > 0) {
+        buflen = BufferLength;
+        mtxt = malloc(buflen);
+    }
+
+    if (EnvironmentHandle) {
+        ret = ESAPI_EnvError(EnvironmentHandle, RecNumber, (SQLCHAR *)qstr_ansi,
+                             NativeError, (SQLCHAR *)mtxt, buflen, &tlen, 0);
+    } else if (ConnectionHandle) {
+        ret = ESAPI_ConnectError(ConnectionHandle, RecNumber,
+                                 (SQLCHAR *)qstr_ansi, NativeError,
+                                 (SQLCHAR *)mtxt, buflen, &tlen, 0);
+    } else if (StatementHandle) {
+        ret = ESAPI_StmtError(StatementHandle, RecNumber, (SQLCHAR *)qstr_ansi,
+                              NativeError, (SQLCHAR *)mtxt, buflen, &tlen, 0);
+    } else {
+        ret = SQL_ERROR;
+    }
+
+    if (SQL_SUCCEEDED(ret)) {
+        if (Sqlstate)
+            utf8_to_ucs2(qstr_ansi, -1, Sqlstate, 6);
+        if (mtxt && tlen <= BufferLength) {
+            SQLULEN ulen = utf8_to_ucs2_lf(mtxt, tlen, FALSE, MessageText,
```

Comment:
SQLCHAR inputs aren't guaranteed to be UTF-8. On Windows it depends on the OS codepage for example (usually codepage CP1252). This should likely be a separate issue to fix.

Reply:
:+1: Will log another issue for this